source/aws_imds_client.c

```diff
@@ -955,6 +873,197 @@ static void s_on_retry_token_acquired(
     }
 }
 
+static enum aws_imds_client_copy_token_result s_requester_copy_token_safely(
+    struct aws_imds_client_user_data *user_data) {
+    AWS_FATAL_ASSERT(user_data);
+    struct aws_imds_client *client = user_data->client;
+    enum aws_imds_client_copy_token_result ret = AWS_IMDS_TK_CP_RS_UNEXPECTED_ERROR;
+    aws_mutex_lock(&client->token_lock);
+    switch (client->token_state) {
+        case AWS_IMDS_TK_ST_VALID: {
+            aws_byte_buf_reset(&user_data->token_result, true);
+            struct aws_byte_cursor cursor = aws_byte_cursor_from_buf(&client->cached_token);
+            if (aws_byte_buf_append_dynamic(&user_data->token_result, &cursor)) {
+                ret = AWS_IMDS_TK_CP_RS_UNEXPECTED_ERROR;
+            } else {
+                ret = AWS_IMDS_TK_CP_RS_SUCCESS;
+            }
+            break;
+        }
+
+        case AWS_IMDS_TK_ST_INVALID: {
```

Comment:
I feel like there's some simplifications here that can be made:

(1) The validity of the client's token feels like a client property and so refreshing and updating  (and retrying) it should be coupled to the client itself.  Can token queries not be associated with a request?  In particular, everything could go in the waiting queue and the user data on the token query could be the client itself.  This would let you collapse UPDATING and WAITING_IN_QUEUE states, everything is now WAITING_IN_QUEUE.  

(2) Once token refreshes are uncoupled from an individual request, you could also simplify the logic of s_imds_client_update_token_safely:

list temp_list;

lock
if success
  swap(queue, temp_list)
else
  if retry_token_failure
    swap(queue, temp_list)
   
unlock

iterate(temp_list)
  completion callback(success or failure)

The loop is now out of the lock, and retry is all or nothing since it's now associated with the client.  The loop's logic is simpler since all you do is the completion callback.

Reply:
It might be good to move token request to client itself to uncouple the logic in requester. Here I did it in requester is to make sure even though one token query failed (tmp network issue), other requesters in the queue still get the chance to acquire token.

Reply:
I have changed the token logic to client and removed state machine